src/lib/stream/ctr/ctr.cpp

```diff
@@ -122,6 +122,44 @@ void CTR_BE::cipher(const uint8_t in[], uint8_t out[], size_t length)
    m_pad_pos += length;
    }
 
+void CTR_BE::write_keystream(uint8_t out[], size_t length)
+   {
+   verify_key_set(m_iv.empty() == false);
+
+   const uint8_t* pad_bits = &m_pad[0];
+   const size_t pad_size = m_pad.size();
+
+   if(m_pad_pos > 0)
+      {
+      const size_t avail = pad_size - m_pad_pos;
+      const size_t take = std::min(length, avail);
+      copy_mem(out, pad_bits + m_pad_pos, take);
+      length -= take;
+      out += take;
+      m_pad_pos += take;
+
+      if(take == avail)
+         {
+         add_counter(m_ctr_blocks);
+         m_cipher->encrypt_n(m_counter.data(), m_pad.data(), m_ctr_blocks);
+         m_pad_pos = 0;
+         }
+      }
+
+   while(length >= pad_size)
+      {
+      copy_mem(out, pad_bits, pad_size);
+      length -= pad_size;
+      out += pad_size;
+
+      add_counter(m_ctr_blocks);
+      m_cipher->encrypt_n(m_counter.data(), m_pad.data(), m_ctr_blocks);
+      }
+
+   copy_mem(out, pad_bits, length);
+   m_pad_pos += length;
+   }
+
```

Comment:
Maybe worthwhile to avoid the code duplication with `CTR_BE::cipher()`? Like so: https://github.com/reneme/botan/commit/3725095adeede3a83477d5d48c76e23c72e0c65c

I'd bet that the compiler inlines that and (hopefully) makes it a zero-cost abstraction.

Reply:
You're right this would work *but* I don't want to do it because it eliminates the possibility of a nice optimization which is missing here in CTR but was already done for SHAKE in 5b5baed264eb1e371cb69983b0e2585454b3b52a

For encrypt the data flow is

buffer <- process(internal_state)
output <- output xor buffer

But for keystream we have:

buffer <- process(internal_state)
output <- buffer

but the buffer isn't doing anything for us. So for SHAKE in the main loop what we do is instead is

output <- process(internal_state)

skipping a memory copy.

I added a todo for CTR mentioning we should do the same. It actually will make a much bigger difference for CTR, because for SHAKE it seems to save ~~ .22 cycles per byte. But CTR + AES-128 (with AES-NI) runs already at about 1.25 cycles per byte - bringing that down to ~1 cpb would make for a large performance improvement.

Reply:
Mhm, I wasn't happy about this additional copy as well. Fair enough.

Reply:
Added a somewhat faster `write_keystream` for CTR as described above in 5aaf48e313bd173e6f6704eaa298fccb923362e2, now down to .78 cpb with large buffers.

Reply:
Agreed. Trying to incorporate that into my proposed abstraction will result in too much added complexity to save a duplicated lines.